app/v2/services/ObjectStoreService.scala

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2023 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package v2.services
+
+import cats.data.EitherT
+import com.google.inject.ImplementedBy
+import play.api.Logging
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.objectstore.client.ObjectSummaryWithMd5
+import uk.gov.hmrc.objectstore.client.play.PlayObjectStoreClientEither
+import uk.gov.hmrc.objectstore.client.Path
+import v2.models.MessageId
+import v2.models.MovementId
+import v2.models.errors.ObjectStoreError
+import v2.models.responses.UpscanResponse.DownloadUrl
+
+import java.net.URL
+import java.time.Clock
+import java.time.OffsetDateTime
+import java.time.ZoneOffset
+import java.time.format.DateTimeFormatter
+import javax.inject.Inject
+import javax.inject.Singleton
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.util.Try
+import scala.util.control.NonFatal
+
+@ImplementedBy(classOf[ObjectStoreServiceImpl])
+trait ObjectStoreService {
+
+  def addMessage(upscanUrl: DownloadUrl, movementId: MovementId, messageId: MessageId)(implicit
+    hc: HeaderCarrier,
+    ec: ExecutionContext
+  ): EitherT[Future, ObjectStoreError, ObjectSummaryWithMd5]
+
+}
+
+@Singleton
+class ObjectStoreServiceImpl @Inject() (clock: Clock, client: PlayObjectStoreClientEither) extends ObjectStoreService with Logging {
+
+  private val dateTimeFormatter = DateTimeFormatter.ofPattern("yyyyMMdd-HHmmss").withZone(ZoneOffset.UTC)
+
+  override def addMessage(downloadUrl: DownloadUrl, movementId: MovementId, messageId: MessageId)(implicit
+    hc: HeaderCarrier,
+    ec: ExecutionContext
+  ): EitherT[Future, ObjectStoreError, ObjectSummaryWithMd5] =
+    EitherT {
+      val formattedDateTime = dateTimeFormatter.format(OffsetDateTime.ofInstant(clock.instant, ZoneOffset.UTC))
+
+      (for {
+        url <- Future.fromTry(Try(new URL(downloadUrl.value)))
+        response <- client
+          .uploadFromUrl(
+            from = url,
+            to = Path.Directory("common-transit-convention-traders").file(s"${movementId.value}-${messageId.value}-$formattedDateTime.xml")
```

Comment:
Please "common-transit-convention-traders" from here as that will be added automatically while generating object store uri.
Upload object from url
osClient.uploadFromUrl(
  from            = new URL("https://fus-outbound.s3.eu-west-2.amazonaws.com/81fb03f5-195d-422a-91ab-460939045846"),
  to              = Path.File(Path.Directory("my-folder"), "sample.pdf"),
  retentionPeriod = RetentionPeriod.OneWeek, // defaults to 'object-store.default-retention-period' configuration
  contentType     = Some("text/csv"), // defaults to None which results in the contentType 'application/octet-stream'
  contentMd5      = Some(Md5Hash("a3c2f1e38701bd2c7b54ebd7b1cd0dbc")), // defaults to None
  owner           = "my-service" // defaults to 'appName' configuration
) // returns Future[ObjectSummaryWithMd5]
The above code will download a file from https://fus-outbound.s3.eu-west-2.amazonaws.com/81fb03f5-195d-422a-91ab-460939045846 to **/my-service/**my-folder/sample.pdf **my-service** is going to be **common-transit-convention-traders** for our case.

Reply:
To make it clearer:

`to` should be to `movements/<movementid>/<filename>`

`owner` should be set to `common-transit-convention-traders` (that would be the default in this case, but it will  make it clear that's the owning service).

Reply:
Also, the retention period should be 7 years.

Reply:
@drnaylor-oc @sonisn1988  The `to`  parameter takes in the following case class. So would the directory be `movements/<movementid>`? 

 case class File(directory: Directory, fileName: String)

Reply:
Yes

Reply:
Thanks guys